File: app/src/main/java/com/example/dictionaryapp/database/DBHelper.java
package com.example.dictionaryapp.database;

import android.content.Context;
import android.os.Build;

import androidx.annotation.RequiresApi;

import com.example.dictionaryapp.TagsHelper;
import com.example.dictionaryapp.deinflector.Deinflector;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class DBHelper {

    private AppDatabase db;
    private DictDao dao;
    private List<String> disabledDicts;
    private boolean shouldDeconj;
    private boolean bilingualFirst;
    private JSONObject inflectorJSON;
    Deinflector deinflector;
    private Context mContext;

    public DBHelper(Context context, List<String> disabledDicts, boolean shouldDeconj, boolean bilingualFirst, JSONObject inflectorJson) {
        this.disabledDicts = disabledDicts;
        mContext = context;
        this.shouldDeconj = shouldDeconj;
        this.bilingualFirst = bilingualFirst;
        this.inflectorJSON = inflectorJson;
        db = AppDatabase.buildDatabase(context);
        dao = db.dictDao();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public List<DictEntry> lookup(String query) throws JSONException {
        System.out.println(this.disabledDicts);
        List<String> possible = normalizeWord(query);
        List<DictEntry> entries = new ArrayList<>();
        if (possible.isEmpty()){
            possible.add(query);
        }
        for (String p : possible) {
            List<DictEntry> res;
            if (isAllKana(p)) {
                String convertedToHiragana;
                if (!allHiragana(p)){
                    convertedToHiragana = katakanaToHiragana(p);
                }
                else {
                    convertedToHiragana = p;
                }
                res = dao.searchEntryReading(convertedToHiragana, this.disabledDicts);
                if (res.isEmpty()){
                    res = dao.searchEntryByKanji(p, this.disabledDicts);
                }
            } else {
                System.out.println("Doing kanji search...");
                res = dao.searchEntryByKanji(p, this.disabledDicts);
            }
            System.out.println(res);
            entries.addAll(res);
        }
        if (this.bilingualFirst) {
            Collections.sort(entries, Collections.reverseOrder());
        } else {
            Collections.sort(entries);
        }
        return entries;

    }

    // helper japanese methods
    public boolean isAllKana(String word) {

        for (int x = 0; x < word.length(); x++) {
            char c = word.charAt(x);
            if ((c > '\u3040' && c < '\u309F') || (c > '\u30A0' && c < '\u30FF')) {
                continue;
            } else {
                return false;
            }
        }
        return true;
    }

    public List<String> normalizeWord(String word) throws JSONException {
        if (this.shouldDeconj) {
            return deconjugateWord(word.trim());
        } else {
            List<String> res = new ArrayList<>();
            res.add(word);
            return res;
        }
    }

    public List<String> deconjugateWord(String word) throws JSONException {
        List<String> results = new ArrayList<>();
        deinflector = new Deinflector(inflectorJSON);
        JSONArray res = deinflector.deinflect(word);
        for (int x = 0; x < res.length(); x++) {
            results.add(res.getJSONObject(x).getString("term"));
        }
        return results;
    }

    public String katakanaToHiragana(String katakanaWord){
        StringBuilder out = new StringBuilder();
        for (int i = 0; i < katakanaWord.length(); i++) {
            out.append(toHiragana(katakanaWord.charAt(i)));
        }
        return out.toString();
    }

    public static char toHiragana(char c) {
        if (isFullWidthKatakana(c)) {
            return (char) (c - 0x60);
        } else if (isHalfWidthKatakana(c)) {
            return (char) (c - 0xcf25);
        }
        return c;
    }

    public static boolean isHalfWidthKatakana(char c) {
        return (('\uff66' <= c) && (c <= '\uff9d'));
    }

    public static boolean isFullWidthKatakana(char c) {
        return (('\u30a1' <= c) && (c <= '\u30fe'));
    }

    public static boolean isHiragana(char c) {
        return (('\u3041' <= c) && (c <= '\u309e'));
    }

    public static boolean allHiragana(String word){
        for (char x : word.toCharArray()){
            if (!isHiragana(x)){
                return false;
            }
        }
        return true;
    }

    public static JSONArray getSplittedTags(DictEntry entry, Context mContext) throws IOException, JSONException {
        TagsHelper helper = new TagsHelper(mContext);
        String[] splitted = entry.getTags().split("\\s+");
        JSONArray info = new JSONArray();
        for (String tag : splitted) {
            if (!tag.isEmpty()) {
                JSONArray temp = new JSONArray();
                String[] tagInfo = helper.getFullTag(tag);
                temp.put(tag);
                temp.put(tagInfo[0]);
                temp.put(tagInfo[1]);
                info.put(temp);
            }
        }
        return info;
    }


}

File: app/src/main/java/com/kamui/rin/database/DBHelper.kt
package com.kamui.rin.database

import android.content.Context
import android.os.Build
import androidx.annotation.RequiresApi
import com.kamui.rin.Tag
import com.kamui.rin.TagsHelper
import com.kamui.rin.deinflector.Deinflector
import java.util.*
import kotlin.collections.ArrayList

class DBHelper(
    mContext: Context,
    private val disabledDicts: List<String>,
    private val shouldDeconj: Boolean,
    private val bilingualFirst: Boolean,
    deinflectionText: String,
) {
    private val db: AppDatabase = AppDatabase.buildDatabase(mContext)
    private val dao: DictDao = db.dictDao()
    private var deinflector: Deinflector = Deinflector(deinflectionText)

    fun lookup(query: String): List<DictEntry> {
        val possibleVariations = normalizeWord(query).toMutableList()
        val entries: MutableList<DictEntry> = ArrayList()
        if (possibleVariations.isEmpty()) {
            possibleVariations.add(query)
        }

        for (variation in possibleVariations) {
            var results: List<DictEntry>
            if (isAllKana(variation)) {
                val convertedToHiragana: String = if (!allHiragana(variation)) {
                    katakanaToHiragana(variation)
                } else {
                    variation
                }
                results = dao.searchEntryReading(convertedToHiragana, disabledDicts)
                if (results.isEmpty()) {
                    results = dao.searchEntryByKanji(variation, disabledDicts)
                }
            } else {
                results = dao.searchEntryByKanji(variation, disabledDicts)
            }
            entries.addAll(results)
        }

        if (bilingualFirst) {
            Collections.sort(entries, Collections.reverseOrder<Any>())
        } else {
            entries.sort()
        }
        return entries
    }

    private fun normalizeWord(word: String): List<String> {
        return if (shouldDeconj) {
            deconjugateWord(word.trim { it <= ' ' })
        } else {
            mutableListOf(word)
        }
    }

    private fun deconjugateWord(word: String): List<String> {
        return deinflector.deinflect(word).map { d -> d.term }
    }
}

fun toHiragana(c: Char): Char {
    if (isFullWidthKatakana(c)) {
        return (c.toInt() - 0x60).toChar()
    } else if (isHalfWidthKatakana(c)) {
        return (c.toInt() - 0xcf25).toChar()
    }
    return c
}

fun isHalfWidthKatakana(c: Char): Boolean {
    return c in '\uff66'..'\uff9d'
}

fun isFullWidthKatakana(c: Char): Boolean {
    return c in '\u30a1'..'\u30fe'
}

fun isHiragana(c: Char): Boolean {
    return c in '\u3041'..'\u309e'
}

fun allHiragana(word: String): Boolean {
    for (x in word.toCharArray()) {
        if (!isHiragana(x)) {
            return false
        }
    }
    return true
}

fun isAllKana(word: String): Boolean {
    for (element in word) {
        return if (element in 'ぁ'..'ゞ' || element in 'ァ'..'ヾ') {
            continue
        } else {
            false
        }
    }
    return true
}

fun katakanaToHiragana(katakanaWord: String): String {
    return katakanaWord.map { c -> toHiragana(c) }.toString()
}

@RequiresApi(Build.VERSION_CODES.KITKAT)
fun getTagsFromSplitted(entry: DictEntry, mContext: Context): List<Tag> {
    val helper = TagsHelper(mContext)
    val splitted: List<String> = entry.tags.split("\\s+")
    return splitted.map { w -> helper.getTagFromName(w) }
}